File: bench/infer.py
"""NumPy equivalent of infer.hk: y = relu(x @ W + b), argmax per row.

This is the honest comparison for the inference claim. NumPy's matmul is
already C (and usually BLAS, multi-threaded), so this is not "compiled beats
interpreted" — it is Halka's generated C against a tuned numerical library.
"""

import time

import numpy as np

ROWS = 4096
FEATS = 512
CLASSES = 32

i = np.arange(FEATS * CLASSES, dtype=np.float64)
w = (0.001 * ((i % 7) - 3.0)).reshape(CLASSES, FEATS)

b = 0.01 * np.arange(CLASSES, dtype=np.float64)

j = np.arange(ROWS * FEATS, dtype=np.float64)
x = (0.5 * ((j % 13) - 6.0)).reshape(ROWS, FEATS)

t0 = time.perf_counter()
y = np.maximum(x @ w.T + b, 0.0)
total = int(y.argmax(axis=1).sum())
ms = (time.perf_counter() - t0) * 1000.0

print(f"infer checksum {total}")
# Reported separately because the wall-clock figure for this script is
# almost entirely `import numpy`, not arithmetic.
print(f"compute {ms:.0f} ms")

File: bench/infer.c
/* Hand-written C equivalent of infer.hk: y = relu(x @ W + b), argmax per row. */
#include <stdio.h>
#include <stdlib.h>
#include <time.h>

#define ROWS 4096
#define FEATS 512
#define CLASSES 32

static double dot(const double *x, const double *w, long xoff, long woff, long n) {
  double acc = 0.0;
  for (long j = 0; j < n; j++) acc += x[xoff + j] * w[woff + j];
  return acc;
}

int main(void) {
  double *w = malloc(sizeof(double) * FEATS * CLASSES);
  double *b = malloc(sizeof(double) * CLASSES);
  double *x = malloc(sizeof(double) * (size_t)ROWS * FEATS);
  if (!w || !b || !x) return 1;

  for (long i = 0; i < (long)FEATS * CLASSES; i++) w[i] = 0.001 * ((double)(i % 7) - 3.0);
  for (long c = 0; c < CLASSES; c++) b[c] = 0.01 * (double)c;
  for (long i = 0; i < (long)ROWS * FEATS; i++) x[i] = 0.5 * ((double)(i % 13) - 6.0);

  clock_t t0 = clock();
  long long total = 0;
  for (long r = 0; r < ROWS; r++) {
    long best = 0;
    double bestv = -1000000.0;
    for (long c = 0; c < CLASSES; c++) {
      double v = dot(x, w, r * FEATS, c * FEATS, FEATS) + b[c];
      if (v < 0.0) v = 0.0;
      if (v > bestv) { bestv = v; best = c; }
    }
    total += best;
  }
  double ms = 1000.0 * (double)(clock() - t0) / CLOCKS_PER_SEC;
  printf("infer checksum %lld\n", total);
  printf("compute %.0f ms\n", ms);
  free(w); free(b); free(x);
  return 0;
}
